---
title: "Introduction to GUSMap (intercross populations)"
author: "Timothy P. Bilton"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Introduction to GUSMap for intercross populations}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: "../inst/REFERENCES.bib"
biblio-style: "apalike"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set( collapse = TRUE, comment = "#>", eval=TRUE)
library(GUSMap)
```

In this tutorial, we will describe how to use GUSMap to perform linkage mapping and explain some of the functionality available. For this tutorial, we will use a F2 or intercross population as implemented in the `IC` class.

For this tutorial, we will use a simulated data set of a full-sib family (F1 population) consisting of 100 offspring, 3 chromosomes and 1000 SNPs to illustrate GUSMap. The file location to the VCF file (and associated pedigree file) for the data set can be extracted using the `simDS` function.

```{r simDS, eval = T}
vcffile <- simDS() # extract filename
vcffile            # filename stored in object vcffile
```

Some important up front comments regarding GUSMap:

1. Having high read depth on the parents (e.g., sequencing more than once) can be hugely helpful to correctly infer the segregation type of more SNPs. However, for the progeny, it is of more value (and cost-efficient) to have more individuals in the family at lower depth than sequence fewer at higher depth.
2. GUSMap is designed for low coverage sequencing data which means that there is no need to pre-filter the data in terms of read depth. Furthermore, GUSMap includes standard filtering steps in terms of minor allele frequency (MAF), missing data and a segregation test (including a few other filters) which means one only needs to input a raw Variant Call Format (VCF) file obtained after the SNP calling processes into GUSMap.

In addition, some terms that we will use throughout this tutorial are:

* Maternal informative (MI) SNPs: SNPs where the maternal parent is heterozygous and the paternal parent is homozygous.
* Paternal informative (PI) SNPs: SNPs where the paternal parent is heterozygous and the maternal maternal is homozygous.
* Both informative (BI) SNPs: SNPs where both parents are heterozygous.

More details of this terminology and the mathematics underpinning the estimation of recombination fractions in GUSMap can be found in @bilton2018genetics1.

## Loading Data

At present, GUSMap can read in data stored in a Variant Call Format (VCF) file, provided that there is some form of allelic depth information (e.g., the number of reads for the reference and alternate alleles). This is done using the `VCFtoRA` function which takes in the name of a VCF file and converts it into what we call a reference/alternate (RA) file.

```{r loadData, eval=T}
# convert VCF file to an RA file
rafile <- VCFtoRA(infilename = vcffile$vcf, direct = "./", makePed=TRUE)
```

The arguments of the `VCFtoRA` function are:

1. `infilename`: The name of the VCF file. 
2. `direct`: Specifies which directory to write the RA file to (relative to the current working directory).
3. `makePed`: Specifies whether or not to initialize a pedigree file. Pedigree files are important for creating a full-sib family object and will be discussed later. 

Currently, `VCFtoRA` requires one of the fields **AD**, **AO and RO**, or **DP4** to be present in the VCF file to extract allelic depth information. One other thing to note is that `VCFtoRA` returns the file location of the created RA file.
```{r rafile, eval=T}
rafile # file path of RA file
```

The RA file created is a tab-delimited file with columns, **CHROM** (the chromosome name taken from the "#CHROM" column in the VCF file), **POS** (the position of the SNP taken from the "POS" column in the VCF file), and **SAMPLES** which consists of the sampleIDs used in the VCF (columns 10 and above in the VCF file). For example, the first five columns and first four rows of the RA file for the simulated data set are:

|CHROM	|POS	   |P1_1  |P2_1 |P1_2 |
|:------|:-------|:-----|:----|:----|
|1	    |866499	 |25,0  |0,2  |9,0  |
|1	    |2070050 |8,0   |4,9  |9,0  |
|1	    |2232468 |1,0	  |0,1  |2,0  |
|1	    |3459263 |12,13 |0,0  |0,1  |

The entry in the first row and the third column is "25,0" which means that 25 reads for the reference allele has been observed and none for the alternate allele. In addition, the fourth row of the third column is "12,13" which menas that 12 reads reads for the reference allele and 13 reads for the alternate allele has been observed at the SNP on chromosome 1 and position 3459263 in individual P1_1. 

An RA file can then be loaded into R using the `readRA` function.
```{r readRA, eval=T}
RAdata <- readRA(rafile = rafile, sampthres = 0.01, excsamp = NULL)
```

The arguments of the `readRA` function are:

1. `rafile`: Name of the RA file to be read in.
2. `sampthres`: Specifies the minimum sample depth of an individual before it is removed.
3. `excsamp`: Specifies IDs of any samples in the RA data set to be discarded (for problematic samples to be removed). The input for `excsamp` must correspond to the sample IDs in the RA file.

The RA data is now stored as an `RA` object.
```{r class_RAdata, eval=T}
class(RAdata)
```

Summary information of the RA object can be displayed by printing the object.
```{r RAdata, eval=T}
RAdata
```

## Linkage mapping with GUSMap

In this section, we will discuss linkage mapping using GUSMap. The main steps to the process are:

1. Create a F2 (intercross) populations and perform appropriate filtering
2. Constructing linkage groups
3. Ordering markers across each linkage group
4. Computing maps for the ordered linkage groups

In the following sections, we shall discuss how each of these steps are performed in GUSMap.

### Constructing a full-sib family

Construction of a intercross (F2) population object is achieved using the `makeIC` function.
```{r makeFS, eval=T}
progeny = RAdata$extractVar("indID")$indID[-c(1:4)]
mySpecies <- makeIC(RAobj = RAdata, samID = progeny,
                      filter = list(MAF = 0.05, MISS = 0.5,
                        BIN = 100, DEPTH = 5, PVALUE = 0.01, MAXDEPTH=500))
```

The `makeFS` function takes an `RA` object and pedigree file, creates a intercross population and performs filtering on the data. 

For this tutorial, we use the existing pedigree file stored in the package. The file name of this pedigree file is stored in the object `vcffile$ped`. More information regarding the format of the pedigree file can be found by looking at the package documentation for the `makeFS` function which can be accessed by:
```{r help_makeIC, eval=F}
?`$makeIC`
```

There are six different types of filtering available in GUSMap:

* Minor allele frequency (`MAF`): SNPs are removed if the MAF is below the threshold since in full-sib families, the expected allele frequency are 0.5 for BI SNPs and 0.75 (or 0.25) for PI and MI SNPs.
* Percentage of missing data (`MISS`): SNPs are removed if the number of individuals without a single read for a given SNP is less than the threshold.
* SNP binning distance (`BIN`): SNPs are binned together if the distance (in base pairs) between them is less than the threshold value. One SNP is then selected at random from each bin. This filtering is to ensure that there is only a single SNP on each sequence.
* Parental depth (`DEPTH`): SNPs are removed if the read depth of each parent is below the threshold value. This filtering is performed since low read depth can result in parental heterozygous genotypes to be seen as homozygous resulting in incorrectly inferred segregation type.  
* Segregation test (`PVALUE`): A segregation test is performed to removed any SNPs in which the segregation type has been incorrectly inferred (e.g., parental genotypes called wrongly).
* Maximum average SNP depth (`MAXDEPTH`): SNPs with an average depth above 500 are discarded. SNPs with really high depth compared to the rest of the SNPs often are repeative regions or simply probamatic. It is often best to discard these SNPs. 

Again, more details of these various filter types is available in the package documentation of the `makeIC` function.

### Linkage Groups


##### Computing 2-point recombination fraction estimates

Linkage group formation in GUSMap is based on 2-point recombination fraction estimates and associated LOD scores between all pairs of SNPs. These are computed using the function `$rf_2pt`. 
```{r 2pt_rf}
mySpecies$rf_2pt(nClust = 3)
```

The `$rf_2pt` function is parallelized to reduce computational time, where the `nClust` argument specifies how many cores to use in the parallelization. Be careful not to set this to more than what is available on your computer. Note: this step can take some time, especially if there are a large number of SNPs in your data set. 

Once the 2-point recombination fraction estimates and associated LOD scores have been computed, we can plot the resulting recombination fraction estimates on a heatmap.
```{r plotChr}
# plot 2-point rf matrix for BI and MI SNPs
mySpecies$plotChr(mat = "rf", lmai=0.5)
```

Heatmaps are helpful for examining genome assemblies if the SNPs where called using a reference genome. For our example, we see that the middle section of chromosome 2 should actually be with chromosome 3, and that the last quarter of chromosome 1 needs to be inverted (that is the marker order need to be reversed). Furthermore, there are a number of SNPs that are clearly on the wrong chromosome and should be mapped to one of the other chromosomes. For *de novo* assemblies, one can still plot the 2-point recombination fraction estimates although the resulting plot should look fairly random and so is usually not very informative.

One can also plot the matrix of 2-point LOD scores.
```{r plotChr2}
# plot 2-point LOD matrix for BI, PI and MI SNPs
mySpecies$plotChr(mat = "LOD", lmai=0.5)
```

In the above code, the `parent` argument has been set to `"both"` which means that all the SNPs will be plotted. The resulting plot looks a somewhat patchy. This is because the LOD scores between PI and MI SNPs are zero as there is no information regarding recombination between these SNPs. The same also applies to the matrix of 2-point recombination fraction estimates.

##### Forming linkage groups

Using our 2-point recombination fraction estimates and LOD scores, we can proceed to forming linkage groups. 
```{r createLG}
mySpecies$createLG(LODthres = 10, nComp = 10)
```

The performance of the linkage grouping algorithm is controlled by the parameters `LODthres` and `nComp`. Details of these arguments and how the linkage groups are formed is explained in the package documentation at:
```{r help_createLG, eval=F}
?`$createLG`
```

Once the linkage groups have been formed, it is a good idea to look how well the grouping has performed by examining the heatmap of the 2-point recombination fraction estimates with the SNPs grouped according to the linkage groups. This can be done using the `$plotLG` function.
```{r plotLG_mat}
mySpecies$plotLG(interactive = F, LG = NULL)
```

The heatmap shows that there are seven linkage groups. We would expect there to be three linkage groups as there are 3 chromosomes. The heatmap suggests that linkage groups 3, 4, 5 and 7 are actually one single linkage group due to the high recombination fraction estimates between linkage groups. It is also possible that linkage group 6 is actually part of linkage group 1, however there is less evidence of this from the 2-point recombination fraction heatmap. 

Note:

* Setting `interactive=TRUE` for the `$plotLG` function results in an interactive plot being plotted. 
Interactive plots make interrogating the recombination fraction (or LOD) plot and finding problematic SNPs easier. We have set this to `FALSE` 
in this tutorial to keep the size of this file small. In your session, try producing the interactive plot.
* At present, interactive plots generally work on a small number of SNPs (usually no more than 1000), so depending on 
the size of your linkage groups, you may need to plot each linkage group separately. You can use the `LG` argument to plot specific linkage groups.


##### Editing Linkage Groups

There are number of functions in GUSMap for editing linkage groups. Ideally, one should obtain as many linkage groups as the number of chromosomes the species is known to have. However, in practice, the number of linkage groups may be more, or less, than the number of chromosomes depending on the LOD threshold used and whether there is enough linkage between SNPs on the same chromosome to group all the SNPs located on the same chromosome together in the same linkage group. 

For our simulated data, we had seven linkage groups. From the heatmaps above, we see that linkage groups 3, 4, 5 and 7 appear part of the same linkage group. We can merge these linkage groups using the `$mergeLG` function.
```{r mergeLG}
mySpecies$mergeLG(LG = c(3,4,5,7))
mySpecies$print(what = "LG")
```

Note that after merging linkage groups, the linkage group which was previous denoted 6 is now denoted by 4. We could also merge this linkage group with linkage group 1 but since we are not confident based on the heatmap, we decide to remove this linkage groups, which can be done using the `$removeLG` function.

```{r removeLG}
mySpecies$removeLG(LG = 4)
mySpecies$print(what = "LG")
```

Alternatively, one could have also have removed the linkage group by removing all the SNPs present in the linkage group using the `$removeSNP` function.
```{r removeSNP, eval=FALSE}
mySpecies$removeSNP(snps = c(36,27,32,28,131))
```

Note that the `$removeSNP` function is more useful for removing SNP(s) from a linkage group if they appear problematic without deleting the entire linkage group. This will be demonstrated later.

**Note:** At this stage there is no undo functionality in GUSMap. So, once you have removed a SNP or linkage group, or merge two linkage groups, you cannot reverse what you have done. However, in the case where you want to undo edits to your linkage group, then the best thing is to recreate the linkage groups using the `$createLG` function and proceed as you were doing before (and hence it helps to save you code in a script).

In practice, our choice for the LOD threshold previously might have been too high to begin with. Thus, we can also attempt to obtain the optimal number of linkage groups by trying different values for `LODthres` in the `$createLG` function.
```{r createLG2, eval=FALSE}
mySpecies$createLG(LODthres = 8)
```

As we can see, this LOD threshold obtains the desired number of linkage groups in one good. In practice, however, there are likely to be lots small linkage groups (in addition to large linkage groups that correspond to the linkage groups) that either need to be merged with other linkage groups or removed using the functions described eariler.  

### Ordering Linkage Groups

Once we are happy with linkage groups we have formed, we can proceed to order the SNPs in each linkage group. In GUSMap, marker ordering is performed using the multidimensional scaling (MDS) approach as described by @preedy2016tag and is implemented in the `$orderLG` function.
```{r orderLG, , fig.width=8, fig.height=6}
set.seed(67584)
mySpecies$orderLG(mapfun = "haldane", weight = "LOD2", ndim = 30) 
```

The two plots produced by the `mySpecies$orderLG` function are:

1. The first two dimensions of the components of the MDS.
2. The heatmap of the 2-point recombination fraction for all SNPs on the linkage group ordered according to the MDS algorithm.

These plots are helpful for examining how well the algorithm has ordered the SNPs. Details of the MDS algorithm are given in @preedy2016tag and information for the arguments `mapfun` and `weight` can be found in the package documentation (The default arguments tend to give the best results). 

```{r help_orderLG, eval=F}
?`$orderLG`
```

### Compute Linkage Maps

Now that we have finished ordering and tidying up the combined linkage groups, we proceed to compute the linkage maps using the `$computeMap` function.
```{r computeMap}
mySpecies$computeMap(chrom = NULL, nThreads = 3, method="EM")
```

The `$computeMap` function has a number of arguments but the main ones are `chrom` which specifies which linkage group to compute 
linkage maps for and `nThreads` which specifies the number of threads to use in the computation of the likelihood (using OpenMP if available).
The resulting output gives the overall map distances and sequencing error estimates for each chromosome (in addition to the 
linkage group information).

GUSMap have a couple of functions for quickly examining the linkage maps computed. The first function is to plot the linkage maps.
```{r plotLM}
mySpecies$plotLM(LG = NULL, fun = "haldane", col=c("cyan","lightblue","blue"))
```

The arguments for the `$plotLM` function are:

* `LG`: The indices of the linkage groups to be plotted.
* `fun`: The mapping function to use. Currently only the Haldane, Kosambi and Morgan mapping functions are available.
* `col`: A vector of colors for the linkage maps.

In addition, we can also examine the order of the SNPs across the linkage groups relative to the original assembly.
```{r plotSyn, fig.width=7, fig.height=7}
mySpecies$plotSyn()
```

From the synteny plot above, we see that linkage group 1 corresponds to chromosome 3, linkage group 2 to chromosome 2 and linkage group 3 to chromosome 1. Furthermore, the plot shows that the linkage mapping analysis has moved the middle section of chromosome 2 to chromosome 3 (linkage group 1) and that it has inverted the SNP order at the end of chromosome 1 (linkage group 3).

Finally, one can extract the results of the linkage analysis in GUSMap by writing the results to a file using the `$writeLM` function.
```{r writeLG, eval=F}
mySpecies$writeLM(file = "mySpecies", inferGeno = TRUE)
```

This is allows users to produce their own custom plots in whatever software they are familiar with. The `inferGeno` argument indicates whether to compute the most likely haplotype for each chromosome in an individual and to append to the file. For details of the information given in the file, see the package documentation.
```{r help_writeLM, eval=F}
?`$writeLM`
```


# References


